core/l1_fee_context.go

```diff
@@ -0,0 +1,97 @@
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var big10 = big.NewInt(10)
+
+var (
+	L1BaseFeeSlot = common.BigToHash(big.NewInt(2))
+	OverheadSlot  = common.BigToHash(big.NewInt(3))
+	ScalarSlot    = common.BigToHash(big.NewInt(4))
+	DecimalsSlot  = common.BigToHash(big.NewInt(5))
+)
+
+// calculateL1GasUsed returns the gas used to include the transaction data in
+// the calldata on L1
+func calculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	var zeroes uint64
+	var ones uint64
+	for _, byt := range data {
+		if byt == 0 {
+			zeroes++
+		} else {
+			ones++
+		}
+	}
+
+	zeroesGas := zeroes * params.TxDataZeroGas
+	onesGas := (ones + 68) * params.TxDataNonZeroGasEIP2028
+	l1Gas := new(big.Int).SetUint64(zeroesGas + onesGas)
+	return new(big.Int).Add(l1Gas, overhead)
+}
+
+// L1FeeContext includes all the context necessary to calculate the cost of
+// including the transaction in L1
+type L1FeeContext struct {
+	BaseFee  *big.Int
+	Overhead *big.Int
+	Scalar   *big.Int
+	Decimals *big.Int
+}
+
+// NewL1FeeContext returns a context for calculating L1 fee cst
+func NewL1FeeContext(cfg *params.ChainConfig, statedb *state.StateDB) *L1FeeContext {
+	// TODO: unpack values after #2596
+	// see: https://github.com/ethereum-optimism/optimism/pull/2596
+	l1BaseFee := statedb.GetState(cfg.Optimism.L1Block, L1BaseFeeSlot).Big()
+	overhead := statedb.GetState(cfg.Optimism.GasPriceOracle, OverheadSlot).Big()
+	scalar := statedb.GetState(cfg.Optimism.GasPriceOracle, ScalarSlot).Big()
+	decimals := statedb.GetState(cfg.Optimism.GasPriceOracle, DecimalsSlot).Big()
+
+	return &L1FeeContext{
+		BaseFee:  l1BaseFee,
+		Overhead: overhead,
+		Scalar:   scalar,
+		Decimals: decimals,
+	}
+}
+
+// L1Cost returns the L1 fee cost.
+// This depends on the chainconfig because gas costs
+// can change over time
+func L1Cost(tx *types.Transaction, ctx *L1FeeContext) *big.Int {
+	var rlp bytes.Buffer
+	if err := tx.EncodeRLP(&rlp); err != nil {
+		panic(err)
+	}
+	divisor := new(big.Int).Exp(big10, ctx.Decimals, nil)
```

Comment:
the divisor can be lifted in the L1FeeContext b/c it does not change unless decimals changes.